include/seqan3/alignment/data_structure/scoring_scheme_base.hpp

```diff
@@ -0,0 +1,220 @@
+#include <range/v3/algorithm/copy.hpp>
+
+#include <seqan3/alphabet/concept.hpp>
+#include <seqan3/core/detail/strong_type.hpp>
+#include <seqan3/core/concept/cereal.hpp>
+
+#if SEQAN3_WITH_CEREAL
+#include <cereal/types/array.hpp>
+#endif // SEQAN3_WITH_CEREAL
+
+namespace seqan3
+{
+
+/*!\brief A strong type of underlying type `int8_t` that represents the score two matching characters.
+ * \ingroup data_structure
+ * \see scoring_scheme_base::set_scheme_simple
+ */
+struct match_score : detail::strong_type<int8_t, match_score, detail::strong_type_skill::convert>
+{
+     using detail::strong_type<int8_t, match_score, detail::strong_type_skill::convert>::strong_type;
+};
+
+/*!\brief A strong type of underlying type `int8_t` that represents the score two different characters.
+ * \ingroup data_structure
+ * \see scoring_scheme_base::set_scheme_simple
+ */
+struct mismatch_score : detail::strong_type<int8_t, mismatch_score, detail::strong_type_skill::convert>
+{
+     using detail::strong_type<int8_t, mismatch_score, detail::strong_type_skill::convert>::strong_type;
+};
+
+/*!\brief A CRTP base class for scoring schemes.
+ * \tparam  derived_t The derived type.
+ * \tparam alphabet_t Type of the largest target alphabet.
+ * \ingroup data_structure
+ *
+ * \details
+ *
+ * This type is never used directly, instead use seqan3::nucleotide_scoring_scheme or seqan3::aminoacid_scoring_scheme.
+ */
+template <typename derived_t, alphabet_concept alphabet>
+class scoring_scheme_base
+{
+public:
+    //!\brief Type of the score values.
+    using score_t = int8_t;
+    //!\brief Size type that can hold the dimension of the matrix (i.e. size of the alphabet).
+    using matrix_size_t = uint8_t;
+    //!\brief Size of the matrix dimensions (i.e. size of the alphabet).
+    static constexpr matrix_size_t matrix_size = alphabet_size_v<alphabet>;
+    //!\brief Type of the internal matrix (a two-dimensional array).
+    using matrix_t = std::array<std::array<score_t, matrix_size>, matrix_size>;
+
+private:
+    //!\brief Befriend derived_t so it can instantiate.
+    friend derived_t;
+
+    //!\publicsection
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr scoring_scheme_base(scoring_scheme_base const &) noexcept = default;
+    constexpr scoring_scheme_base(scoring_scheme_base &&) noexcept = default;
+    constexpr scoring_scheme_base & operator=(scoring_scheme_base const &) noexcept = default;
+    constexpr scoring_scheme_base & operator=(scoring_scheme_base &&) noexcept = default;
+
+    //!\brief The default constructor (delegates to set_scheme_hamming()).
+    constexpr scoring_scheme_base() noexcept
+    {
+        set_scheme_hamming();
+    }
+
+    /*!\brief Constructor for the simple scheme (delegates to set_scheme_simple()).
+     * \copydetails set_scheme_simple()
+     */
+    constexpr scoring_scheme_base(match_score const ms, mismatch_score const mms) noexcept
+    {
+        set_scheme_simple(ms, mms);
```

Comment:
We could add an enum `unknown_policy` to treat matching against unknowns differently. 
Either `::matches_unknown`, `::matches_all`, `::matches_nothing`.

Reply:
I don't know what you mean by this and I didn't get it from your files either.